doc/developer/design/20210504_manage_external_state.md

```diff
@@ -0,0 +1,203 @@
+
+* Is there any external state that *needs* to be created in the `dataflow` layer? (Where it's not possible to create it
+earlier?)
+* Same question as above, but for destroying state.
+
+## Issues raised during discussion
+
+This design document was discussing on 5/5/2021, in the External Integrations weekly team sync. The following thoughts and
+issues were raised.
+
+### Source instance state
+
+The Coordinator may spin up a source than in turn spins up multiple source instances. If each of those source instances
+requires its own state, and we want to store external state information in `coord`, we will need to enable one of two
+things:
+1. Have a way for source instances to communicate back to the Coordinator, which could cause race conditions.
+1. Pre-create any state needed by a source instance in the Coordinator, and then pass that information to each source
+instance via some `Connector` information.
+
+The team agreed that the second approach was preferable, even if it updated the interface of some `dataflow` functions.
+However, storing source instance state introduces a bit of persistence complexity. Whatever instance-specific state we
+store in the Catalog must be stable across versions or versioned.
+
+Note: the same may be true for sink instances of some type, but was not discussed.
+
+### Source instance error paths
+
+It was brought up that if there were an issue with a source or sink at the `dataflow` layer, the dataflow operators would
+not be empowered to clean up state themselves or communicate this back to the Coordinator. While this is a valid objection,
+it matches current behavior. If there is an issue with a dataflow, the user will have to manually `DROP` it. Moving forward,
+as we provide more useful errors, we may want to revisit this.
+
+### pg_replication_slots table in Materialize
+
+It was proposed that we create a new `pg_replication_slots` table in Materialize's Catalog in order to track the external
+state of Postgres sources. (This suggestion is Postgres source specific, breaking a bit with the goal of this design doc, but
+helping us think of more concrete solutions.) `pg_replication_slots` will be like Materialize's timestamps table.
+
+### Add a new CatalogItem variant
+
+It was proposed that, instead of the proposal in this document, that we create a new `CatalogItem` type to track and store
+information relevant to external state. Unfortunately, `CatalogItem` is a specific type of resolveable object, and information
+about external state does not match that pattern. We will not be moving forward with this alternate proposal.
+
+### Difference between this approach and desired Timestamper refactors
+
+It was brought up that we want to move the Timestamper from `coord` and into `dataflow`, which is the opposite of what this
+document proposes for external state. The reason being that each BYO source must share the same Timestamper, and it would
+be preferable for each source to get its own thread.
```

Comment:
Having a centralized timestamper thread was a totally valid design, and possibly even a desirable one. It didn't pan out not because having a centralized timestamper thread is inherently a bad idea but because 1) Kafka's API makes this painful and 2) the thread wasn't designed in a scalable way.